# ML Assignment 5 – Random Forest, Naive Bayes, SVM

## Crowdfunding Campaign Success Prediction

# Machine Learning Model Building Steps:-

## Step - 1. Data Preprocessing:
### 1.1.	Load the Dataset: 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv(r"https://raw.githubusercontent.com/ArchanaInsights/Datasets/refs/heads/main/crowdfunding_campaign.csv")
df.head()

CampaignID  GoalAmount  RaisedAmount  DurationDays  NumBackers Category  \
0  CID_00000       76159  73518.028955            15        1903     Film   
1  CID_00001       69394  60068.683288            49         469      Art   
2  CID_00002       18652  22958.153470            58         148     Film   
3  CID_00003       19487  24229.048390            39         780     Film   
4  CID_00004       26734  17984.630311            68        1622    Games   

  LaunchMonth    Country Currency  OwnerExperience VideoIncluded  \
0   September  Australia      GBP               11           Yes   
1     October  Australia      GBP                7           Yes   
2    February        USA      EUR               17            No   
3     October    Germany      AUD               10            No   
4   September  Australia      AUD               12           Yes   

   SocialMediaPresence  NumUpdates  IsSuccessful  
0                41758           3             0  
1                79112          48             0  
2                 3942           9             1  
3                22047          35             1  
4                72568          17             0

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 14 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   CampaignID           100000 non-null  object 
 1   GoalAmount           100000 non-null  int64  
 2   RaisedAmount         100000 non-null  float64
 3   DurationDays         100000 non-null  int64  
 4   NumBackers           100000 non-null  int64  
 5   Category             100000 non-null  object 
 6   LaunchMonth          100000 non-null  object 
 7   Country              100000 non-null  object 
 8   Currency             100000 non-null  object 
 9   OwnerExperience      100000 non-null  int64  
 10  VideoIncluded        100000 non-null  object 
 11  SocialMediaPresence  100000 non-null  int64  
 12  NumUpdates           100000 non-null  int64  
 13  IsSuccessful         100000 non-null  int64  
dtypes: float64(1), int64(7), object(6)
memory usage: 10.7+ MB


In [4]:
df.describe()

GoalAmount   RaisedAmount   DurationDays     NumBackers  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean    50538.001310   50598.121803      52.047370    1003.347340   
std     28613.822592   33255.277493      21.653238     574.485153   
min      1000.000000     528.632349      15.000000      10.000000   
25%     25724.750000   23396.500017      33.000000     505.000000   
50%     50706.000000   45980.674476      52.000000    1003.000000   
75%     75316.000000   72790.183158      71.000000    1502.000000   
max     99999.000000  149238.361559      89.000000    1999.000000   

       OwnerExperience  SocialMediaPresence     NumUpdates   IsSuccessful  
count    100000.000000        100000.000000  100000.000000  100000.000000  
mean          9.489130         50077.301010      24.483230       0.500940  
std           5.763538         28767.521522      14.413833       0.500002  
min           0.000000           100.000000       0.000000       0.000000  
25%           4.000000         25276.750000      12.000000       0.000000  
50%           9.000000         50024.500000      25.000000       1.000000  
75%          14.000000         74934.250000      37.000000       1.000000  
max          19.000000         99999.000000      49.000000       1.000000

In [5]:
df.shape

(100000, 14)

## 1.2.	Handle Missing Values: 

In [6]:
df.isnull().sum()

CampaignID             0
GoalAmount             0
RaisedAmount           0
DurationDays           0
NumBackers             0
Category               0
LaunchMonth            0
Country                0
Currency               0
OwnerExperience        0
VideoIncluded          0
SocialMediaPresence    0
NumUpdates             0
IsSuccessful           0
dtype: int64

- There is no null values

## 1.3.Encode Categorical Features: 

In [7]:
from sklearn.preprocessing import LabelEncoder 

le = LabelEncoder()
model_df = df.copy()
model_df["VideoIncluded"] =le.fit_transform(model_df["VideoIncluded"])
model_df.head()

CampaignID  GoalAmount  RaisedAmount  DurationDays  NumBackers Category  \
0  CID_00000       76159  73518.028955            15        1903     Film   
1  CID_00001       69394  60068.683288            49         469      Art   
2  CID_00002       18652  22958.153470            58         148     Film   
3  CID_00003       19487  24229.048390            39         780     Film   
4  CID_00004       26734  17984.630311            68        1622    Games   

  LaunchMonth    Country Currency  OwnerExperience  VideoIncluded  \
0   September  Australia      GBP               11              1   
1     October  Australia      GBP                7              1   
2    February        USA      EUR               17              0   
3     October    Germany      AUD               10              0   
4   September  Australia      AUD               12              1   

   SocialMediaPresence  NumUpdates  IsSuccessful  
0                41758           3             0  
1                79112          48             0  
2                 3942           9             1  
3                22047          35             1  
4                72568          17             0

#### Using One_Hot_encoding or get_dummies for Multiple unique Category values columns.[Category,LaunchMonth,Country and Currency]

In [8]:
cat_col=["Category","LaunchMonth","Country","Currency"]
model_df = pd.get_dummies(model_df,columns=cat_col,drop_first=True)
model_df.head()

CampaignID  GoalAmount  RaisedAmount  DurationDays  NumBackers  \
0  CID_00000       76159  73518.028955            15        1903   
1  CID_00001       69394  60068.683288            49         469   
2  CID_00002       18652  22958.153470            58         148   
3  CID_00003       19487  24229.048390            39         780   
4  CID_00004       26734  17984.630311            68        1622   

   OwnerExperience  VideoIncluded  SocialMediaPresence  NumUpdates  \
0               11              1                41758           3   
1                7              1                79112          48   
2               17              0                 3942           9   
3               10              0                22047          35   
4               12              1                72568          17   

   IsSuccessful  ...  LaunchMonth_October  LaunchMonth_September  \
0             0  ...                False                   True   
1             0  ...                 True                  False   
2             1  ...                False                  False   
3             1  ...                 True                  False   
4             0  ...                False                   True   

   Country_Canada  Country_Germany  Country_UK  Country_USA  Currency_CAD  \
0           False            False       False        False         False   
1           False            False       False        False         False   
2           False            False       False         True         False   
3           False             True       False        False         False   
4           False            False       False        False         False   

   Currency_EUR  Currency_GBP  Currency_USD  
0         False          True         False  
1         False          True         False  
2          True         False         False  
3         False         False         False  
4         False         False         False  

[5 rows x 33 columns]

In [9]:
model_df.shape

(100000, 33)

## 1.4.	Feature Selection: 

In [10]:
X = model_df.drop(["CampaignID","IsSuccessful"],axis=1) 
X.head()

GoalAmount  RaisedAmount  DurationDays  NumBackers  OwnerExperience  \
0       76159  73518.028955            15        1903               11   
1       69394  60068.683288            49         469                7   
2       18652  22958.153470            58         148               17   
3       19487  24229.048390            39         780               10   
4       26734  17984.630311            68        1622               12   

   VideoIncluded  SocialMediaPresence  NumUpdates  Category_Film  \
0              1                41758           3           True   
1              1                79112          48          False   
2              0                 3942           9           True   
3              0                22047          35           True   
4              1                72568          17          False   

   Category_Games  ...  LaunchMonth_October  LaunchMonth_September  \
0           False  ...                False                   True   
1           False  ...                 True                  False   
2           False  ...                False                  False   
3           False  ...                 True                  False   
4            True  ...                False                   True   

   Country_Canada  Country_Germany  Country_UK  Country_USA  Currency_CAD  \
0           False            False       False        False         False   
1           False            False       False        False         False   
2           False            False       False         True         False   
3           False             True       False        False         False   
4           False            False       False        False         False   

   Currency_EUR  Currency_GBP  Currency_USD  
0         False          True         False  
1         False          True         False  
2          True         False         False  
3         False         False         False  
4         False         False         False  

[5 rows x 31 columns]

In [11]:
y = model_df["IsSuccessful"]
y.head()

0    0
1    0
2    1
3    1
4    0
Name: IsSuccessful, dtype: int64

## 1.5.	Data Splitting: 

In [12]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.25,random_state=42)

## 1.6.	Feature Scaling

In [13]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaler = scaler.fit_transform(X_train)
X_test_scaler = scaler.transform(X_test)

# Step-2.	Random Forest - Model Building and Evaluation:

### 2.1.	Model Building: 

In [14]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

n_estimators = [50, 100, 150]
max_depth = [5, 10, 15]   

best_score = 0
print("Random Forest Accuracy Results:\n")

for n in n_estimators:
    for d in max_depth:
        Rf_model = RandomForestClassifier(n_estimators=n, max_depth=d, random_state=42)
        Rf_model.fit(X_train_scaler, y_train)
        y_pred_Rf = Rf_model.predict(X_test_scaler)
        acc_Rf = accuracy_score(y_test, y_pred_Rf)
        print(f"n_estimators: {n}, max_depth: {d}, Accuracy: {acc_Rf:.4f}")
       
        # store the best score
        if  acc_Rf > best_score:
            best_score = acc_Rf
            
        
# Model Evaluation
print("Best Accuracy:", best_score)
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred_Rf))
print("\nClassification Report:\n", classification_report(y_test, y_pred_Rf))

Random Forest Accuracy Results:

n_estimators: 50, max_depth: 5, Accuracy: 0.7074
n_estimators: 50, max_depth: 10, Accuracy: 0.8141
n_estimators: 50, max_depth: 15, Accuracy: 0.8517
n_estimators: 100, max_depth: 5, Accuracy: 0.7066
n_estimators: 100, max_depth: 10, Accuracy: 0.8036
n_estimators: 100, max_depth: 15, Accuracy: 0.8540
n_estimators: 150, max_depth: 5, Accuracy: 0.7028
n_estimators: 150, max_depth: 10, Accuracy: 0.7993
n_estimators: 150, max_depth: 15, Accuracy: 0.8559
Best Accuracy: 0.85592

Confusion Matrix:
 [[12284   174]
 [ 3428  9114]]

Classification Report:
               precision    recall  f1-score   support

           0       0.78      0.99      0.87     12458
           1       0.98      0.73      0.83     12542

    accuracy                           0.86     25000
   macro avg       0.88      0.86      0.85     25000
weighted avg       0.88      0.86      0.85     25000



# Step - 3.	Naive Bayes - Model Building and Evaluation:

### 3.1.	Model Building:
- We use Gaussian Naive Bayes Model (When the data follow Continuous Values)

In [15]:
from sklearn.naive_bayes import MultinomialNB, GaussianNB, BernoulliNB

Gnb_model= GaussianNB()
Gnb_model.fit(X_train_scaler, y_train)

GaussianNB()

### 3.2.	Model Evaluation: 

In [16]:
y_pred_Gnb = Gnb_model.predict(X_test_scaler)
acc_Gnb = accuracy_score(y_test, y_pred_Gnb)
print("Accuracy Score :",acc_Gnb)
print("\nConfusion Matrix:\n",confusion_matrix(y_test, y_pred_Gnb))
print("\nClassification Report:\n",classification_report(y_test, y_pred_Gnb))

Accuracy Score : 0.66116

Confusion Matrix:
 [[10003  2455]
 [ 6016  6526]]

Classification Report:
               precision    recall  f1-score   support

           0       0.62      0.80      0.70     12458
           1       0.73      0.52      0.61     12542

    accuracy                           0.66     25000
   macro avg       0.68      0.66      0.65     25000
weighted avg       0.68      0.66      0.65     25000



# Step-4.	Support Vector Machine (SVM) - Model Building and Evaluation:

### 4.1.	Model Building: 

In [17]:
from sklearn.svm import SVC

SVM_model = SVC(kernel="linear", random_state=42)
SVM_model.fit(X_train_scaler, y_train)
y_pred_SVM = SVM_model.predict(X_test_scaler)
acc_SVM_linear = accuracy_score(y_test, y_pred_SVM)

#4.2 Model Evaluation
print("Linear_SVM Accuracy:", acc_SVM_linear)
print("\nConfusion Matrix:\n",confusion_matrix(y_test, y_pred_SVM))
print("\nClassification Report:\n",classification_report(y_test, y_pred_SVM))

Linear_SVM Accuracy: 0.99708

Confusion Matrix:
 [[12440    18]
 [   55 12487]]

Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00     12458
           1       1.00      1.00      1.00     12542

    accuracy                           1.00     25000
   macro avg       1.00      1.00      1.00     25000
weighted avg       1.00      1.00      1.00     25000



In [11]:
# Model running time is Very slow 


"""SVM_model_r = SVC(kernel="rbf", random_state=42)
SVM_model_r.fit(X_train_scaler, y_train)
y_pred_SVM_r = SVM_model_r.predict(X_test_scaler)
acc_SVM_rbf = accuracy_score(y_test, y_pred_SVM_r)

print("Rbf_SVM Accuracy:", acc_SVM_rbf)
print("\nConfusion Matrix:\n",confusion_matrix(y_test, y_pred_SVM_r))
print("\nClassification Report:\n",classification_report(y_test, y_pred_SVM_r))"""

'SVM_model_r = SVC(kernel="rbf", random_state=42)\nSVM_model_r.fit(X_train_scaler, y_train)\ny_pred_SVM_r = SVM_model_r.predict(X_test_scaler)\nacc_SVM_rbf = accuracy_score(y_test, y_pred_SVM_r)\n\nprint("Rbf_SVM Accuracy:", acc_SVM_rbf)\nprint("\nConfusion Matrix:\n",confusion_matrix(y_test, y_pred_SVM_r))\nprint("\nClassification Report:\n",classification_report(y_test, y_pred_SVM_r))'

# 5.	Comparison and Analysis:

### 5.1.	Compare the accuracy scores of the three models.

- Random Forest Accuracy : 0.85592               
- GaussianNB Accuracy : 0.66116                 
- Support Vector Machine(SVM) Accuracy : 0.99708                

- **Best Accuracy:** Support Vector Machine(SVM) Accuracy : 0.99708 

### 5.2.	Discuss which model performed best and why, based on the accuracy scores and other relevant factors.

- **Best Model:** Linear SVM (Accuracy: 0.99708)
- Highest accuracy among all three models
- Likely indicates the dataset is linearly separable after scaling, and SVM works great with high-dimensional spaces.

### 5.3.	Reflect on the strengths and weaknesses of each model in the context of this dataset.

#### 1.Random Forest:
**Strenghs:**
- Handles both linear and non-linear relationships well.
- Feature importance is easy to interpret.
    
**Weaknesses:**
- Slower to train (especially with many trees or deep trees).

#### 2.Gaussian Naive Bayes:
**Strenghs:**
- Very fast to train and predict.
- Works well with large datasets

**Weaknesses:**
- Assumes feature independence, which is rarely true in real data.
- Lower accuracy due to that assumption.

#### 3.SVM 
**Strenghs:**
- Extremely high accuracy in this case.
- Works beautifully with scaled numeric data.
    
**Weaknesses:**
- The training time for the model is very high.
